pandas/_libs/algos.pyx

```diff
@@ -793,216 +797,288 @@ ctypedef fused rank_t:
-    na_option="keep",
     bint pct=False,
+    na_option="keep",
 ):
     """
     Fast NaN-friendly version of ``scipy.stats.rankdata``.
+
+    Parameters
+    ----------
+    in_arr : array of rank_t values to be ranked
+    labels : array containing unique label for each group, with its ordering
+        matching up to the corresponding record in `values`. If not called
+        from a groupby operation, will be an array of 0's
+    ties_method : {'average', 'min', 'max', 'first', 'dense'}, default
+        'average'
+        * average: average rank of group
+        * min: lowest rank in group
+        * max: highest rank in group
+        * first: ranks assigned in order they appear in the array
+        * dense: like 'min', but rank always increases by 1 between groups
+    ascending : boolean, default True
+        False for ranks by high (1) to low (N)
+        na_option : {'keep', 'top', 'bottom'}, default 'keep'
+    pct : boolean, default False
+        Compute percentage rank of data within each group
+    na_option : {'keep', 'top', 'bottom'}, default 'keep'
+        * keep: leave NA values where they are
+        * top: smallest rank if ascending
+        * bottom: smallest rank if descending
     """
     cdef:
-        Py_ssize_t i, j, n, dups = 0, total_tie_count = 0, non_na_idx = 0
-        ndarray[rank_t] sorted_data, values
-        ndarray[float64_t] ranks
-        ndarray[int64_t] argsorted
-        ndarray[uint8_t, cast=True] sorted_mask
-        rank_t val, nan_value
-        float64_t sum_ranks = 0
-        int tiebreak = 0
-        bint keep_na = False
-        bint isnan, condition
-        float64_t count = 0.0
+        TiebreakEnumType tiebreak
+        Py_ssize_t i, j, N, grp_start=0, dups=0, sum_ranks=0
+        Py_ssize_t grp_vals_seen=1, grp_na_count=0, grp_tie_count=0
+        ndarray[int64_t, ndim=1] _as
+        ndarray[float64_t, ndim=1] grp_sizes, out
+        ndarray[rank_t, ndim=1] masked_vals
+        ndarray[uint8_t, ndim=1] mask
+        bint keep_na, at_end, next_val_diff, check_labels
+        rank_t nan_fill_val
 
     tiebreak = tiebreakers[ties_method]
+    keep_na = na_option == 'keep'
 
-    if rank_t is float64_t:
-        values = np.asarray(in_arr).copy()
-    elif rank_t is object:
-        values = np.array(in_arr, copy=True)
+    N = len(in_arr)
+    out = np.empty(N)
+    grp_sizes = np.ones(N)
+    # If all 0 labels, can short-circuit later label
+    # comparisons
+    check_labels = np.count_nonzero(labels) != 0
 
-        if values.dtype != np.object_:
-            values = values.astype('O')
-    else:
-        values = np.asarray(in_arr).copy()
-
-    keep_na = na_option == 'keep'
+    # Copy values into new array in order to fill missing data
+    # with mask, without obfuscating location of missing data
+    # in values array
+    masked_vals = np.array(in_arr, copy=True)
+    if rank_t is object and masked_vals.dtype != np.object_:
+        masked_vals = masked_vals.astype('O')
```

Comment:
you can *just* do this for object types (e.g. do an else and then)
`mask_vals = in_array.copy()` since .astype always copies

Reply:
Thanks